src/test/java/com/thoughtworks/qdox/GenericsTest.java

```diff
@@ -139,7 +140,25 @@ public void testShouldUnderstandMethodTypeParameters() {
         builder.addSource(new StringReader(source));
         assertEquals("Bar", builder.getClassByName("Bar").getName());
     }
-    
+
+    public void testShouldUnderstandAnnotationsOnTypeParameters() {
+        String source = "" +
+                "public class Bar {" +
+                "    public static Collection<String> foo() { }" +
+                "    public static Collection<java.lang.String> foo2() { }" +
+                "    public static Collection<@Annot String> foo3() { }" +
+                "    public static Collection<java.lang.@Annot String> foo4() { }" +
```

Comment:
This line looks wrong to me. I assume you `public static Collection<@Annot java.lang.String> foo4() { }`.
If so, I have a patch matching your requirements and the JLS

Reply:
You're right. It does ***look*** wrong. Unfortunately that's how it's supposed to be.

`private Collection<@NotEmpty java.lang.String> test;` does not compile, but
`private Collection<java.lang.@NotEmpty String> test;` does.

That's also what the spec says:

> PackageName . {Annotation} TypeIdentifier [TypeArguments]

Reply:
https://github.com/paul-hammant/qdox/commit/92e615c8880b2285adb053b01f4f1a84ea715883 contains the proper support for TypeVariable. The other testcase requires support for ClassType (already available as comment, but not implemented and not updated with the latest specs.)
Do you want to give it a try?

Reply:
I think that covers my situation, but it also contains the invalid case of `public static Collection<@Annot java.lang.String> foo4() { }` which should be `public static Collection<java.lang.@Annot String> foo4() { }`

Reply:
I can't find good examples or explanations, but does `java.lang` belong to `@Annot` or to `String`? Where is this specified?

Reply:
So it belongs to String. https://github.com/paul-hammant/qdox/commit/8074fba3b02ab1945a1ee49a201829e9a40f55db contains the code to support that too.
I noticed that for JLS11 I should go through the whole file again to get the terms in line again.